MennaMohie/assignment3/Solution.cpp

```diff
@@ -0,0 +1,172 @@
+#include<iostream>
+#include<string>
+#include<vector>
+#include<map>
+#include<unordered_map>
+#include<set>
+#include"Testing.h" //test macros + overloaded insertion operator for sets
+using namespace std;
+
+enum checkType { wordCheck, prefixCheck };
+
+const int gridCols = 3;
+
+class Trie;
+
+class TrieNode
+{
+	char character;
+	bool terminates = false;
+	unordered_map<char, TrieNode*> children;
+
+	void addWord(string word)
+	{
+		int wordSize = word.length();
+
+		if (wordSize == 0)
+			return;
+
+		char firstChar = word[0];
+		TrieNode *child = children[firstChar];
+		if (child == nullptr)
+		{
+			child = new TrieNode(firstChar);
+			children[firstChar] = child;
+		}
+		if(wordSize>1)
+			child->addWord(word.substr(1, wordSize - 1));
+		else
+			child->terminates = true;
+	}
+
+	bool hasChild(char c)
+	{
+		return children.find(c) != children.end();
+	}
+
+public:
+	TrieNode() {}
+	TrieNode(char c) :character(c){}
+	
+	friend class Trie;	
+};
+
+class Trie
+{
+	TrieNode *root;
+
+public:
+
+	Trie(vector<string> words)
+	{
+		root = new TrieNode();
+		for (string word : words)
+		{
+			root->addWord(word);
+		}
+	}
+
+	bool isWord(string word, checkType type = wordCheck)
+	{
+		if (word.length() == 0)
+			return false;
+
+		TrieNode *node = root;
+
+		for (char c : word)
+		{
+			if (!node->hasChild(c))
+				return false;
+			node = node->children[c];
+
+			//if we're doing a wordCheck and we've reached the last letter in the word we need to check its terminates boolean
+			if (type == wordCheck && c == word[word.length() - 1] && node->terminates == false)
+				return false;
+		}
+
+		return true;
+	}
+
+	bool isPrefix(string prefix)
+	{
+		return isWord(prefix, prefixCheck);
+	}
+};
+
+bool formWords(string word, Trie dictionary, char grid[][gridCols], int rows, int currentRow, int currentCol, vector<string> &words)
```

Comment:
On each call, you may be making a copy of all parameters, which can have performance implications.
Also, is a bit restrictive to allow boards of only 3 columns. May be better to have a `vector<vector<char>>` or `vector<string>`